src/main/java/com/google/cloud/flink/bigquery/BigQueryDirectDataWriterContext.java

```diff
@@ -0,0 +1,219 @@
+import com.google.cloud.flink.bigquery.common.BigQueryDirectDataWriteHelper;
+import com.google.cloud.flink.bigquery.common.BigQueryDirectWriterCommitMessageContext;
+import com.google.cloud.flink.bigquery.common.DataWriterContext;
+import com.google.cloud.flink.bigquery.common.UserAgentHeaderProvider;
+import com.google.cloud.flink.bigquery.common.WriterCommitMessageContext;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Descriptors;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import net.sf.jsqlparser.JSQLParserException;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.StructuredType;
+import org.apache.flink.table.types.logical.StructuredType.StructuredAttribute;
+import org.apache.flink.types.Row;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BigQueryDirectDataWriterContext implements DataWriterContext<Row> {
+
+  final Logger logger = LoggerFactory.getLogger(BigQueryDirectDataWriterContext.class);
+
+  final String tablePath;
+  final StructuredType flinkSchema;
+  final Descriptors.Descriptor schemaDescriptor;
+  static BigQuery bigquery;
+  private BigQueryDirectDataWriteHelper writeHelper;
+
+  Table srcTable;
+  String projectId;
+  String dataset;
+
+  String table;
+  StandardSQLTypeName type;
+
+  FlinkBigQueryConfig bqconfig;
+
+  public BigQueryDirectDataWriterContext(
+      Table srcTable, String projectId, String dataset, String table) throws JSQLParserException {
+    List<DataType> columnDataTypeList = Arrays.asList(srcTable.getSchema().getFieldDataTypes());
+    List<String> columnNameList = Arrays.asList(srcTable.getSchema().getFieldNames());
+    ArrayList<StructuredAttribute> listOfAttribute = new ArrayList<StructuredAttribute>();
+    for (int i = 0; i < columnNameList.size(); i++) {
+      listOfAttribute.add(
+          new StructuredAttribute(
+              columnNameList.get(i).toString(), columnDataTypeList.get(i).getLogicalType()));
+    }
+    final StructuredType.Builder builder = StructuredType.newBuilder(Row.class);
+    builder.attributes(listOfAttribute);
+    builder.setFinal(true);
+    builder.setInstantiable(true);
+
+    this.projectId = projectId;
+    this.dataset = dataset;
+    this.table = table;
+    flinkSchema = builder.build();
+    this.tablePath = String.format("projects/%s/datasets/%s/tables/%s", projectId, dataset, table);
+    this.srcTable = srcTable;
+    try {
+      this.schemaDescriptor = toDescriptor(flinkSchema);
+    } catch (Descriptors.DescriptorValidationException e) {
+      throw new BigQueryConnectorException.InvalidSchemaException(
+          "Could not convert flink-schema to descriptor object", e);
+    }
+
+    this.bqconfig = BigQueryDynamicTableFactory.getBqConfig();
+
+    BigQueryCredentialsSupplier bigQueryCredentialsSupplier =
```

Comment:
Instead of creating objects like BigQueryCredentialsSupplier, UserAgentHeaderProvider and BigQueryClientFactory here, cant they be injected from where they are called? It will make unit testing much easier and also improve the readability